File: recidiviz/case_triage/case_updates.py
# Recidiviz - a data platform for criminal justice reform
# Copyright (C) 2021 Recidiviz, Inc.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
# =============================================================================
"""Implements interface for querying client_actions_taken."""
from enum import Enum
from typing import Any, Dict, List, Optional

from sqlalchemy.dialects.postgresql import insert
from sqlalchemy.orm import Session

from recidiviz.persistence.database.schema.case_triage.schema import CaseUpdate, ETLOfficer


class CaseUpdateActionType(Enum):
    DISCHARGE_INITIATED = 'DISCHARGE_INITIATED'
    DOWNGRADE_INITIATED = 'DOWNGRADE_INITIATED'

    NOT_ON_CASELOAD = 'NOT_ON_CASELOAD'


class CaseUpdatesInterface:
    """Implements interface for querying case_updates."""

    @staticmethod
    def update_case_for_person(session: Session,
                               officer: ETLOfficer,
                               person_external_id: str,
                               actions: List[CaseUpdateActionType],
                               other_text: Optional[str] = None) -> None:
        """This method updates the case_updates table with the newly provided actions.

        Because the underlying table does not have foreign key constraints, independent
        validation must be provided before calling this method.
        """
        update_metadata: Dict[str, Any] = {'actions': [a.value for a in actions]}
        if other_text:
            update_metadata['otherText'] = other_text

        insert_statement = insert(CaseUpdate).values(
            person_external_id=person_external_id,
            officer_external_id=officer.external_id,
            state_code=officer.state_code,
            update_metadata=update_metadata,
        ).on_conflict_do_update(
            index_elements=['person_external_id', 'officer_external_id', 'state_code'],
            set_={
                'update_metadata': update_metadata,
            },
        )
        session.execute(insert_statement)
        session.commit()

File: frontends/case-triage/src/stores/CaseUpdatesStore/CaseUpdates.ts
// Recidiviz - a data platform for criminal justice reform
// Copyright (C) 2021 Recidiviz, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
// =============================================================================
import { OpportunityType } from "../OpportunityStore/Opportunity";

export enum CaseUpdateActionType {
  // Risk Assessment
  INCORRECT_ASSESSMENT_DATA = "INCORRECT_ASSESSMENT_DATA",
  // Employment
  INCORRECT_EMPLOYMENT_DATA = "INCORRECT_EMPLOYMENT_DATA",
  // Face to face contact
  INCORRECT_CONTACT_DATA = "INCORRECT_CONTACT_DATA",
  // Home visit contact
  INCORRECT_HOME_VISIT_DATA = "INCORRECT_HOME_VISIT_DATA",
  // Supervision Level
  INCORRECT_SUPERVISION_LEVEL_DATA = "INCORRECT_SUPERVISION_LEVEL_DATA",

  NOT_ON_CASELOAD = "NOT_ON_CASELOAD",
  CURRENTLY_IN_CUSTODY = "CURRENTLY_IN_CUSTODY",

  INCORRECT_NEW_TO_CASELOAD_DATA = "INCORRECT_NEW_TO_CASELOAD_DATA",
}

export const isErrorReport = (action: CaseUpdateActionType): boolean => {
  switch (action) {
    case CaseUpdateActionType.INCORRECT_SUPERVISION_LEVEL_DATA:
    case CaseUpdateActionType.INCORRECT_ASSESSMENT_DATA:
    case CaseUpdateActionType.INCORRECT_EMPLOYMENT_DATA:
    case CaseUpdateActionType.INCORRECT_CONTACT_DATA:
    case CaseUpdateActionType.INCORRECT_HOME_VISIT_DATA:
    case CaseUpdateActionType.INCORRECT_NEW_TO_CASELOAD_DATA:
    case CaseUpdateActionType.NOT_ON_CASELOAD:
    case CaseUpdateActionType.CURRENTLY_IN_CUSTODY:
      return true;
    default:
      return false;
  }
};

export enum CaseUpdateStatus {
  IN_PROGRESS = "IN_PROGRESS",
}

export interface CaseUpdate {
  actionTs: string;
  actionType: CaseUpdateActionType;
  comment: string | null;
  status: CaseUpdateStatus;
  updateId?: string;
}

export const CASE_UPDATE_OPPORTUNITY_ASSOCIATION: Record<
  OpportunityType,
  CaseUpdateActionType
> = {
  [OpportunityType.OVERDUE_DOWNGRADE]:
    CaseUpdateActionType.INCORRECT_SUPERVISION_LEVEL_DATA,
  [OpportunityType.EMPLOYMENT]: CaseUpdateActionType.INCORRECT_EMPLOYMENT_DATA,
  [OpportunityType.ASSESSMENT]: CaseUpdateActionType.INCORRECT_ASSESSMENT_DATA,
  [OpportunityType.CONTACT]: CaseUpdateActionType.INCORRECT_CONTACT_DATA,
  [OpportunityType.HOME_VISIT]: CaseUpdateActionType.INCORRECT_HOME_VISIT_DATA,
  [OpportunityType.NEW_TO_CASELOAD]:
    CaseUpdateActionType.INCORRECT_NEW_TO_CASELOAD_DATA,
};

export const ACTION_TITLES: Record<CaseUpdateActionType, string> = {
  [CaseUpdateActionType.INCORRECT_SUPERVISION_LEVEL_DATA]:
    "Incorrect supervision level data",
  [CaseUpdateActionType.INCORRECT_ASSESSMENT_DATA]: "Assessment not needed",
  [CaseUpdateActionType.INCORRECT_EMPLOYMENT_DATA]: "Employment not needed",
  [CaseUpdateActionType.INCORRECT_CONTACT_DATA]: "Contact not needed",
  [CaseUpdateActionType.INCORRECT_HOME_VISIT_DATA]: "Home visit not needed",
  [CaseUpdateActionType.INCORRECT_NEW_TO_CASELOAD_DATA]:
    "Incorrect time on caseload",
  [CaseUpdateActionType.NOT_ON_CASELOAD]: "Not on Caseload",
  [CaseUpdateActionType.CURRENTLY_IN_CUSTODY]: "In Custody",
};
